README.md
# react-native-smart-button

[![npm](https://img.shields.io/npm/v/react-native-smart-button.svg)](https://www.npmjs.com/package/react-native-smart-button)
[![npm](https://img.shields.io/npm/dm/react-native-smart-button.svg)](https://www.npmjs.com/package/react-native-smart-button)
[![npm](https://img.shields.io/npm/dt/react-native-smart-button.svg)](https://www.npmjs.com/package/react-native-smart-button)
[![npm](https://img.shields.io/npm/l/react-native-smart-button.svg)](https://github.com/react-native-component/react-native-smart-button/blob/master/LICENSE)

A smart button for react-native apps, written in JS for cross-platform support.
It works on iOS and Android.

This component is compatible with React Native 0.25 and newer.

Inspired by [react-native-button][6]

## Preview

![react-native-smart-button-preview-ios][5]
![react-native-smart-button-preview-android][8]

## Installation

```
npm install react-native-smart-button --save
```

## Full Demo

see [ReactNativeComponentDemos][0]

## Usage

Install the package from npm with `npm install react-native-smart-button --save`.
Then, require it from your app's JavaScript files with `import Button from 'react-native-smart-button'`.

If you need use badge button, Install the badge from npm with `npm install react-native-smart-badge --save`.
Then, require it from your app's JavaScript files with `import Badge from 'react-native-smart-badge'`.

```js
import React, {Component} from 'react'
import {
    ScrollView,
    View,
    StyleSheet,
    Image,
    Text,
    ActivityIndicator,
    ActivityIndicatorIOS,
    ProgressBarAndroid,
    Platform,
} from 'react-native'

import TimerEnhance from 'react-native-smart-timer-enhance'
import Button from 'react-native-smart-button'
import image_liking from '../images/liking.png'
import Badge from 'react-native-smart-badge'

const styles = StyleSheet.create({
    buttonStyle: {
        margin: 10,
        height: 40,
        backgroundColor: 'red',
        borderRadius: 3,
        borderWidth: StyleSheet.hairlineWidth,
        borderColor: 'red',
        justifyContent: 'center',
    },
    buttonTextStyle: {
        fontSize: 17,
        color: 'white'
    },
    buttonDisabledStyle: {
        backgroundColor: '#DDDDDD',
        borderWidth: 0,
    },
    buttonDisabledTextStyle: {
        color: '#BCBCBC',
    },
});

class AllButton extends Component {

    // 构造
    constructor (props) {
        super(props)
        // 初始状态
        this.state = {
            btn_1_loading: false,
            btn_2_loading: false,
            btn_3_loading: false,
            btn_1_disabled: false,
            btn_2_disabled: false,
            btn_3_disabled: false,
        }
    }

    render () {
        return (
            <ScrollView style={{flex: 1, marginTop: 20 + 44, }}>

                <Button
                    disabled={true}
                    style={styles.buttonStyle}
                    textStyle={styles.buttonTextStyle}
                    disabledStyle={styles.buttonDisabledStyle}
                    disabledTextStyle={styles.buttonDisabledTextStyle}>
                    disabled (按钮禁用)
                </Button>

                <Button
                    style={styles.buttonStyle}
                    textStyle={styles.buttonTextStyle}>
                    opacity all (按钮透明)
                </Button>

                <Button
                    touchableType={Button.constants.touchableTypes.fadeContent}
                    style={styles.buttonStyle}
                    textStyle={styles.buttonTextStyle}>
                    opacity content (内容透明)
                </Button>

                <Button
                    touchableType={Button.constants.touchableTypes.highlight}
                    underlayColor={'#C90000'}
                    style={styles.buttonStyle}
                    textStyle={styles.buttonTextStyle}>
                    highlight (背景高亮)
                </Button>

                <Button
                    touchableType={Button.constants.touchableTypes.blur}
                    style={styles.buttonStyle}
                    textStyle={styles.buttonTextStyle}>
                    blur for ios (模糊阴影)
                </Button>

                <Button
                    loading={this.state.btn_1_loading}
                    disabled={this.state.btn_1_disabled}
                    touchableType={Button.constants.touchableTypes.fade}
                    style={styles.buttonStyle}
                    textStyle={styles.buttonTextStyle}
                    renderLoadingComponent={this._renderLoadingComponent_1}
                    onPress={this._handlePress_1}>
                    loading (加载器)
                </Button>

                <Button
                    loading={this.state.btn_2_loading}
                    disabled={this.state.btn_2_disabled}
                    touchableType={Button.constants.touchableTypes.fadeContent}
                    style={styles.buttonStyle}
                    textStyle={styles.buttonTextStyle}
                    renderLoadingComponent={this._renderLoadingComponent_2}
                    onPress={this._handlePress_2}>
                    loading (加载器+文字)
                </Button>

                <Button
                    touchableType={Button.constants.touchableTypes.highlight}
                    underlayColor={'#C90000'}
                    style={styles.buttonStyle}
                    textStyle={styles.buttonTextStyle}>
                    <Image source={image_liking} style={{width: 40, height: 40, marginRight: 3, }}/>
                    icon (图标)
                </Button>

                <Button
                    touchableType={Button.constants.touchableTypes.blur}
                    style={styles.buttonStyle}
                    textStyle={styles.buttonTextStyle}>
                    badge (徽章)
                    <Badge
                        style={{ backgroundColor: '#00AAEF', marginLeft: 6, }}
                        textStyle={{ color: '#fff', fontSize: 12, }}>
                        8
                    </Badge>
                </Button>

            </ScrollView>
        )
    }

    _handlePress_1 = () => {
        this.setState({
            btn_1_loading: true,
            // btn_1_disabled: true,
        })
        this.setTimeout(() => {
            this.setState({
                btn_1_loading: false,
                // btn_1_disabled: false
            })
        }, 3000)
    }

    _handlePress_2 = () => {
        this.setState({
            btn_2_loading: true,
            // btn_2_disabled: true,
        })
        this.setTimeout(() => {
            this.setState({
                btn_2_loading: false,
                // btn_2_disabled: false
            })
        }, 3000)
    }

    _renderLoadingComponent_1 = () => {
        return this._renderActivityIndicator()
    }

    _renderLoadingComponent_2 = () => {
        return (
            <View style={{flexDirection: 'row', alignItems: 'center'}}>
                {this._renderActivityIndicator()}
                <Text style={{
                    fontSize: 17,
                    color: 'white',
                    fontWeight: 'bold',
                    fontFamily: '.HelveticaNeueInterface-MediumP4',
                }}>loading</Text>
            </View>
        )
    }

    _renderActivityIndicator() {
        return ActivityIndicator ? (
            <ActivityIndicator
                style={{margin: 10,}}
                animating={true}
                color={'#fff'}
                size={'small'}/>
        ) : Platform.OS == 'android' ?
            (
                <ProgressBarAndroid
                    style={{margin: 10,}}
                    color={'#fff'}
                    styleAttr={'Small'}/>

            ) :  (
            <ActivityIndicatorIOS
                style={{margin: 10,}}
                animating={true}
                color={'#fff'}
                size={'small'}/>
        )


    }

}

export default TimerEnhance(AllButton)
```

## Props

Prop                    | Type   | Optional | Default        | Description
----------------------- | ------ | -------- | -------------- | -----------
touchableType           | string | Yes      | constants.fade | determines what kind of style of the button should be when touch is active.
activeOpacity           | number | Yes      | 0.2            | see [react-native documents][1]
underlayColor           | color  | Yes      |                | see [react-native documents][1]
style                   | style  | Yes      |                | see [react-native documents][2]
textStyle               | style  | Yes      |                | see [react-native documents][7]
disabledStyle           | style  | Yes      |                | see [react-native documents][2]
disabledTextStyle       | style  | Yes      |                | see [react-native documents][7]
onPressIn               | func   | Yes      |                | see [react-native documents][3]
onPressOut              | func   | Yes      |                | see [react-native documents][3]
onPress                 | func   | Yes      |                | see [react-native documents][3]
onLayout                | func   | Yes      |                | see [react-native documents][3]
disabled                | bool   | Yes      |                | see [react-native documents][3]
loading                 | bool   | Yes      | false          | determines the loading status of the button
renderLoadingComponent  | func   | Yes      |                | determines render function which return the presentation to replaces the content when loading status of the button is true
shadowOpacity           | number | Yes      | 1              | see [react-native documents][4]
shadowColor             | color  | Yes      | '#fff'         | see [react-native documents][4]

[0]: https://github.com/cyqresig/ReactNativeComponentDemos
[1]: https://facebook.github.io/react-native/docs/touchablehighlight.html
[2]: https://facebook.github.io/react-native/docs/style.html
[3]: https://facebook.github.io/react-native/docs/touchablewithoutfeedback.html#props
[4]: https://facebook.github.io/react-native/docs/shadowproptypesios.html
[5]: http://cyqresig.github.io/img/react-native-smart-button-preview-ios-v1.3.0.gif
[6]: https://github.com/ide/react-native-button
[7]: http://facebook.github.io/react-native/docs/text.html#style
[8]: http://cyqresig.github.io/img/react-native-smart-button-preview-android-v1.3.0.gif


CHANGELOG.md
# 2.0.0

* refactor
* let component controlled
* add eslint

# 1.5.0

* add feature blur.




Blur.js
import React, {
    Component,
    PropTypes,
} from 'react'
import {
    Animated,
    View,
} from 'react-native'

import {
    DEFAULT_SHADOW_COLOR,
    DEFAULT_SHADOW_OPACITY,
    DEFAULT_SHADOW_ANIMATED,
    DEFAULT_SHADOW_DURATION,
} from './constants'

export default class Blur extends Component {

    static defaultProps = {
        shadowColor: DEFAULT_SHADOW_COLOR,
        shadowOpacity: DEFAULT_SHADOW_OPACITY,
        animated: DEFAULT_SHADOW_ANIMATED,
        duration: DEFAULT_SHADOW_DURATION,
    }

    static propTypes = {
        show: PropTypes.bool.isRequired,    // eslint-disable-line
        // textDimension: PropTypes.shape({
        //     width: PropTypes.number.isRequired,
        //     height: PropTypes.number.isRequired,
        // }),
        containerDimension: PropTypes.object.isRequired,    // eslint-disable-line
        shadowOpacity: PropTypes.number,
        shadowColor: PropTypes.string,
        animated: PropTypes.bool,
        duration: PropTypes.number, // eslint-disable-line
    }

    // 构造
    constructor(props) {
        super(props)
        // 初始状态
        this.state = {
            opacity: props.animated ? new Animated.Value(0) : 0,
        }
    }

    componentWillReceiveProps(nextProps) {
        if (!this._boxDimension && nextProps.containerDimension) {
            this._boxDimension = this._getBoxDimension(nextProps.containerDimension)
            this._boxOffset = this._getBoxOffset(this._boxDimension)
            this._shadowOffset = this._getShadowOffset(nextProps.containerDimension, this._boxDimension)
            this._shadowRadius = this._getShadowRadius(this._boxDimension)
        }
        this._toogleBlur(nextProps)
    }

    render() {
        let renderView = null
        if (this._boxDimension) {
            if (this.props.animated) {
                renderView = (
                    <Animated.View
                        style={{
                            opacity: this.state.opacity,
                            position: 'absolute',
                            width: this._boxDimension.width,
                            height: this._boxDimension.height,
                            left: this._boxOffset.x,
                            top: this._boxOffset.y,
                            borderRadius: this._shadowRadius,
                            shadowColor: this.props.shadowColor,
                            shadowOpacity: this.props.shadowOpacity,
                            shadowOffset: this._shadowOffset,
                            shadowRadius: this._shadowRadius,
                        }}/>
                )
            } else {
                console.log(`this.props.animated = `, this.props.animated)
                renderView = (
                    <View
                        style={{
                            opacity: this.state.opacity,
                            position: 'absolute',
                            width: this._boxDimension.width,
                            height: this._boxDimension.height,
                            left: this._boxOffset.x,
                            top: this._boxOffset.y,
                            borderRadius: this._shadowRadius,
                            shadowColor: this.props.shadowColor,
                            shadowOpacity: this.props.shadowOpacity,
                            shadowOffset: this._shadowOffset,
                            shadowRadius: this._shadowRadius,
                        }}/>
                )
            }
        }
        return renderView
    }

    _getBoxDimension(containerDimension) {
        return {
            // width: textDimension.width,
            width: containerDimension.height,
            height: containerDimension.height,
        }
    }

    _getBoxOffset(boxDimension) {
        return {
            x: -boxDimension.width,
            y: -boxDimension.height,
        }
    }

    _getShadowOffset(containerDimension, boxDimension) {
        return {
            width: containerDimension.width / 2 + boxDimension.width / 2,
            height: containerDimension.height / 2 + boxDimension.height / 2,
        }
    }

    _getShadowRadius(boxDimension) {
        return (Math.min(boxDimension.width, boxDimension.height) / 2) * 0.5
    }

    _toogleBlur = (nextProps) => {
        if (this.props.animated) {
            Animated.timing(
                this.state.opacity,
                {
                    toValue: nextProps.show ? 1 : 0,
                    duration: nextProps.duration,
                }
            ).start()
        } else {
            this.setState({
                opacity: nextProps.show ? 1 : 0,
            })
        }
    }
}



blur.js

import React, {
  Component,
  PropTypes,
} from 'react'
import {
  Animated,
  Easing,
  View,
} from 'react-native'


const DEFAULT_SHADOW_COLOR = '#fff'
const DEFAULT_SHADOW_OPACITY = 1
const DEFAULT_SHADOW_ANIMATED = true
const DEFAULT_SHADOW_DURATION = 150

export default class Blur extends Component {

  static defaultProps = {
    shadowColor: DEFAULT_SHADOW_COLOR,
    shadowOpacity: DEFAULT_SHADOW_OPACITY,
    animated: DEFAULT_SHADOW_ANIMATED,
    duration: DEFAULT_SHADOW_DURATION
  }

  static PropTypes = {
    show: PropTypes.bool,
    textDimension: PropTypes.object,
    containerDimension: PropTypes.object,
    shadowOpacity: PropTypes.number,
    shadowColor: PropTypes.color,
    animated: PropTypes.bool,
    duration: PropTypes.number,
  }

  // 构造
    constructor(props) {
      super(props);
      // 初始状态
      this.state = {
        opacity: props.animated ? new Animated.Value(0) : 0,
      };
    }

  componentWillReceiveProps (nextProps) {
    if(this.boxDimension == null) {
      this.boxDimension = this._getBoxDimension(nextProps.textDimension, nextProps.containerDimension)
      this.boxOffset = this._getBoxOffset(this.boxDimension)
      this.shadowOffset = this._getShadowOffset(nextProps.containerDimension, this.boxDimension)
      this.shadowRadius = this._getShadowRadius(this.boxDimension)
    }
    this._toogleBlur(nextProps)
  }

  render() {
      if(this.boxDimension == null) {
        return null
      }
      else {
        return this.props.animated ? (
          <Animated.View
            style={{
              opacity: this.state.opacity,
              position: 'absolute',
              width: this.boxDimension.width,
              height: this.boxDimension.height,
              left: this.boxOffset.x,
              top: this.boxOffset.y,
              borderRadius: this.shadowRadius,
              shadowColor: this.props.shadowColor,
              shadowOpacity: this.props.shadowOpacity,
              shadowOffset: this.shadowOffset,
              shadowRadius: this.shadowRadius,
             }}
          ></Animated.View>
        ) : (
          <View
            style={{
              opacity: this.state.opacity,
              position: 'absolute',
              width: this.boxDimension.width,
              height: this.boxDimension.height,
              left: this.boxOffset.x,
              top: this.boxOffset.y,
              borderRadius: this.shadowRadius,
              shadowColor: this.props.shadowColor,
              shadowOpacity: this.props.shadowOpacity,
              shadowOffset: this.shadowOffset,
              shadowRadius: this.shadowRadius,
             }}
          ></View>
        )
      }
  }

  //componentDidUpdate() {
  //  this._toogleBlur()
  //}

  _getBoxDimension(textDimension, containerDimension) {
    return {
      width: textDimension.width,
      height: containerDimension.height,
    }
  }

  _getBoxOffset(boxDimension) {
    return {
      x: -boxDimension.width,
      y: -boxDimension.height,
    }
  }

  _getShadowOffset(containerDimension, boxDimension) {
    return {
      width: containerDimension.width / 2 + boxDimension.width / 2,
      height: containerDimension.height / 2 + boxDimension.height / 2,
    }
  }

  _getShadowRadius(boxDimension) {
    return Math.min(boxDimension.width, boxDimension.height) / 2
  }

  _toogleBlur = (nextProps) => {
    if(this.props.animated) {
      Animated.timing(
        this.state.opacity,
        {
          toValue: nextProps.show ? 1 : 0,
          duration: nextProps.duration,
        }
      ).start()
    }
    else {
      this.setState({
        opacity: nextProps.show ? 1 : 0,
      })
    }

  }
}



Button.js
/*
 * A smart button for react-native apps
 * https://github.com/react-native-component/react-native-smart-button/
 * Released under the MIT license
 * Copyright (c) 2016 react-native-component <moonsunfall@aliyun.com>
 */

import React, {
    PureComponent,
    PropTypes,
} from 'react'
import {
    View,
    Text,
    TouchableHighlight,
    TouchableOpacity,
    StyleSheet,
} from 'react-native'

import constants, {touchableTypes, SYSTEM_OPACITY, NOOP, EMPTY_OBJECT,} from './constants'
import Blur from './Blur'

const styles = StyleSheet.create({
    text: {
        color: '#007aff',
        fontFamily: '.HelveticaNeueInterface-MediumP4',
        fontSize: 17,
        fontWeight: 'bold',
        textAlign: 'center',
    },
    disabledText: {
        color: '#dcdcdc',
    },
    touchContainer: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
    },
    contentContainer: {
        flexDirection: 'row',
        justifyContent: 'center',
        alignItems: 'center',
    },
})

export default class Button extends PureComponent {

    static constants = constants

    static defaultProps = {
        touchableType: touchableTypes.fade,
        activeOpacity: SYSTEM_OPACITY,
        loading: false,
        disabled: false,
        onPress: NOOP,
        onPressIn: NOOP,
        onPressOut: NOOP,
        onLayout: NOOP,
        renderLoadingComponent: NOOP,
    }

    static propTypes = {
        testID: PropTypes.string,
        touchableType: PropTypes.oneOf([
            touchableTypes.highlight,
            touchableTypes.fadeContent,
            touchableTypes.blur,
            touchableTypes.fade,
        ]),
        activeOpacity: PropTypes.number,
        underlayColor: PropTypes.string,
        style: View.propTypes.style,
        disabledStyle: View.propTypes.style,
        textStyle: Text.propTypes.style,
        disabledTextStyle: Text.propTypes.style,
        onPressIn: PropTypes.func,
        onPressOut: PropTypes.func,
        onPress: PropTypes.func,
        onLayout: PropTypes.func,
        disabled: PropTypes.bool,
        loading: PropTypes.bool,
        renderLoadingComponent: PropTypes.func,
        shadowOpacity: PropTypes.number,
        shadowColor: PropTypes.string,
    }

    constructor(props) {
        super(props)
        this.state = {
            pressIn: false,
            disabled: props.disabled,
            loading: props.loading,
        }
        this._boxDimension = EMPTY_OBJECT
    }

    render() {
        const touchableProps = EMPTY_OBJECT
        const {touchableType,} = this.props

        touchableProps.onPress = this.props.onPress
        touchableProps.onPressIn = this.props.onPressIn
        touchableProps.onPressOut = this.props.onPressOut

        let renderView = null

        switch (touchableType) {
            case touchableTypes.highlight:
                renderView = this._renderHighlightView(touchableProps)
                break
            case touchableTypes.fadeContent:
                renderView = this._renderFadeContentView(touchableProps)
                break
            case touchableTypes.blur: // experimental feature
                renderView = this._renderBlurView(touchableProps)
                break
            case touchableTypes.fade:
            default:
                renderView = this._renderFadeView(touchableProps)
                break
        }
        return renderView
    }

    componentWillReceiveProps(nextProps) {
        const { disabled: nextDisabled, loading: nextLoading, } = nextProps
        const { disabled, loading, } = this.props
        const newState = {}
        if (nextDisabled !== disabled) {
            newState.disabled = nextDisabled
        } else if (nextLoading !== loading) {
            newState.loading = nextLoading
        }
        if (Object.keys(newState).length > 0) {
            this.setState(newState)
        }
    }

    _renderHighlightView(touchableProps) {
        const reassignedProps = {
            ...touchableProps,
            activeOpacity: 1,
            underlayColor: this.props.underlayColor,
        }
        return (
            <TouchableHighlight
                onLayout={this.props.onLayout}
                disabled={this.state.disabled || this.state.loading}
                style={[
                    this.props.style,
                    this.state.disabled ? this.props.disabledStyle : null,
                ]}
                {...reassignedProps}
                testID={this.props.testID}
            >
                {this._renderChildren()}
            </TouchableHighlight>
        )
    }

    _renderFadeContentView(touchableProps) {
        const reassignedProps = {
            ...touchableProps,
            activeOpacity: this.props.activeOpacity,
        }
        return (
            <View
                onLayout={this.props.onLayout}
                style={[
                    this.props.style,
                    this.state.disabled ? this.props.disabledStyle : null,
                ]}>
                <TouchableOpacity
                    disabled={this.state.disabled || this.state.loading}
                    style={[styles.touchContainer,]}
                    {...reassignedProps}
                    testID={this.props.testID}
                >
                    {this._renderChildren()}
                </TouchableOpacity>
            </View>
        )
    }

    _renderBlurView(touchableProps) {
        const reassignedProps = {
            ...touchableProps,
            activeOpacity: this.props.activeOpacity,
            onPressIn: this._onBlurPressIn,
            onPressOut: this._onBlurPressOut,
        }
        return (
            <View
                onLayout={this._handleButtonLayout}
                style={[
                    {
                        position: 'relative',
                        overflow: 'hidden',
                    },
                    this.props.style,
                    this.state.disabled ? this.props.disabledStyle : null,
                ]}>
                <TouchableOpacity
                    disabled={this.state.disabled || this.state.loading}
                    style={styles.touchContainer}
                    {...reassignedProps}
                    testID={this.props.testID}
                >
                    {this._renderChildren()}
                </TouchableOpacity>
                {this._renderBlur()}
            </View>
        )
    }

    _renderFadeView(touchableProps) {
        const reassignedProps = {
            ...touchableProps,
            activeOpacity: this.props.activeOpacity,
        }
        return (
            <TouchableOpacity
                onLayout={this.props.onLayout}
                disabled={this.state.disabled || this.state.loading}
                style={[
                    this.props.style,
                    this.state.disabled ? this.props.disabledStyle : null,
                ]}
                {...reassignedProps}
                testID={this.props.testID}
            >
                {this._renderChildren()}
            </TouchableOpacity>
        )
    }

    _renderBlur() {
        return (
            <Blur
                show={this.state.pressIn}
                // textDimension={this.textDimension}
                containerDimension={this._boxDimension}
                shadowOpacity={this.props.shadowOpacity}
                shadowColor={this.props.shadowColor}
            />
        )
    }

    _renderChildren() {
        let childView = null
        if (this.state.loading) {
            childView = (
                <View style={styles.contentContainer}>
                    {this.props.renderLoadingComponent()}
                </View>
            )
        } else {
            const children = React.Children.map(this.props.children, (child) => {
                if (!React.isValidElement(child)) {
                    return (
                        <Text
                            // onLayout={this._onTextLayout}
                            style={[
                                styles.text,
                                this.props.textStyle,
                                this.state.disabled ? (this.props.disabledTextStyle || styles.disabledText) : null,
                            ]}>
                            {child}
                        </Text>
                    )
                }
                return child
            })
            childView = (
                <View style={styles.contentContainer}>
                    {children}
                </View>
            )
        }
        return childView
    }

    _onBlurPressIn = (e) => {
        this.setState({
            pressIn: true,
        })
        this.props.onPressIn(e)
    }

    _onBlurPressOut = (e) => {
        this.setState({
            pressIn: false,
        })
        this.props.onPressOut(e)
    }

    _handleButtonLayout = (e) => {
        this._boxDimension = {
            width: e.nativeEvent.layout.width,
            height: e.nativeEvent.layout.height,
        }
        this.props.onLayout(e)
    }

    // _onTextLayout = (e) => {
    //    this.textDimension = {
    //      width: e.nativeEvent.layout.width,
    //      height: e.nativeEvent.layout.height,
    //    }
    // }

}


button.js
/*
 * A smart button for react-native apps
 * https://github.com/react-native-component/react-native-smart-button/
 * Released under the MIT license
 * Copyright (c) 2016 react-native-component <moonsunfall@aliyun.com>
 */

import React, {
  Component,
  PropTypes,
} from 'react'
import {
  View,
  Text,
  TouchableHighlight,
  TouchableOpacity,
  TouchableWithoutFeedback,
  StyleSheet,
} from 'react-native'

import Blur from './blur'

const SYSTEM_OPACITY = 0.2
const TOUCH_TYPE = 'opacity'

export default class Button extends Component {

  static defaultProps = {
    touchableType: TOUCH_TYPE,
    activeOpacity: SYSTEM_OPACITY,
    onPress: () => {},
    onPressIn: () => {},
    onPressOut: () => {},
  }

  static PropTypes = {
      testID: PropTypes.string,
      touchableType: PropTypes.string,
      activeOpacity: PropTypes.number,
      underlayColor: PropTypes.color,
      style: View.propTypes.style,
      textStyle: Text.propTypes.style,
      onPressIn: PropTypes.func,
      onPressOut: PropTypes.func,
      onPress: PropTypes.func,
      disabled: PropTypes.bool,
      icon: PropTypes.element,
      badge: PropTypes.element,
      shadowOpacity: PropTypes.number,
      shadowColor: PropTypes.color,
  }

    // 构造
    constructor(props, context) {
      super(props);
      // 初始状态
      this.state = {
         pressIn: false,
      };
    }

  render() {
      let touchableProps = {},
          touchableType = this.props.touchableType
      if(!this.props.disabled) {
        touchableProps.onPress = this.props.onPress
        touchableProps.onPressIn = this.props.onPressIn
        touchableProps.onPressOut = this.props.onPressOut
      }
      switch (touchableType) {
          case 'highlight':
              touchableProps.activeOpacity = 1
              touchableProps.underlayColor = this.props.underlayColor
              return (
                <TouchableHighlight
                  style={[this.props.style]}
                  {...touchableProps}
                  testID={this.props.testID}>
                  <View>
                    {this.props.icon}
                    <Text style={[styles.text, this.props.disabled ? styles.disabledText : null, this.props.textStyle]}>
                      {this.props.children}
                    </Text>
                    {this.props.badge}
                  </View>
                </TouchableHighlight>
              )
           case 'opacityContent':
             touchableProps.activeOpacity = this.props.activeOpacity
             return (
               <View style={this.props.style}>
                 <TouchableOpacity
                   style={[styles.touchContainer]}
                   {...touchableProps}
                   testID={this.props.testID}>
                   {this.props.icon}
                   <Text style={[styles.text, this.props.disabled ? styles.disabledText : null, this.props.textStyle]}>
                     {this.props.children}
                   </Text>
                   {this.props.badge}
                 </TouchableOpacity>
               </View>
             )
           case 'blur': //experimental feature
             touchableProps.activeOpacity = this.props.activeOpacity
             touchableProps.onPressIn = this._onBlurPressIn
             touchableProps.onPressOut = this._onBlurPressOut
             return (
               <View onLayout={this._onButtonLayout} style={[this.props.style, {position: 'relative', overflow: 'hidden',}]}>
                 <TouchableOpacity
                   style={styles.touchContainer}
                   {...touchableProps}
                   testID={this.props.testID}>
                   {this.props.icon}
                   <Text
                     onLayout={this._onTextLayout}
                     style={[styles.text, this.props.disabled ? styles.disabledText : null, this.props.textStyle]}
                   >
                     {this.props.children}
                   </Text>
                   {this.props.badge}
                 </TouchableOpacity>
                 {this._renderBlur()}
               </View>
             )
           case 'opacity':
           default:
             touchableProps.activeOpacity = this.props.activeOpacity
             return (
                 <TouchableOpacity
                   style={[this.props.style]}
                   {...touchableProps}
                   testID={this.props.testID}>
                   {this.props.icon}
                   <Text style={[styles.text, this.props.disabled ? styles.disabledText : null, this.props.textStyle]}>
                     {this.props.children}
                   </Text>
                   {this.props.badge}
                 </TouchableOpacity>
               )
      }
  }

  _renderBlur() {
      return  (
          <Blur
            show={this.state.pressIn}
            textDimension={this.textDimension}
            containerDimension={this.boxDimension}
            shadowOpacity={this.props.shadowOpacity}
            shadowColor={this.props.shadowColor}
          />
      )
  }

  _onBlurPressIn = (e) => {
      this.setState({
        pressIn: true,
      })
      this.props.onPressIn.call(this, e)
  }

  _onBlurPressOut = (e) => {
    this.setState({
      pressIn: false,
    })
    this.props.onPressOut.call(this, e)
  }

  _onButtonLayout = (e) => {
    this.boxDimension = {
      width: e.nativeEvent.layout.width,
      height: e.nativeEvent.layout.height,
    }
  }

  _onTextLayout = (e) => {
     this.textDimension = {
       width: e.nativeEvent.layout.width,
       height: e.nativeEvent.layout.height,
     }
  }

}

const styles = StyleSheet.create({
  text: {
    color: '#007aff',
    fontFamily: '.HelveticaNeueInterface-MediumP4',
    fontSize: 17,
    fontWeight: 'bold',
    textAlign: 'center',
  },
  disabledText: {
    color: '#dcdcdc',
  },
  touchContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center'
  },
})